pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -341,40 +359,70 @@ func updateStatusFromPod(taskRun *v1alpha1.TaskRun, pod *corev1.Pod) {
 	taskRun.Status.PodName = pod.Name
 
 	taskRun.Status.Steps = []v1alpha1.StepState{}
+	completed := len(pod.Status.ContainerStatuses) > 0 && pod.Status.Phase == corev1.PodRunning
+	failure := false
+	sidecarsReady := 0
 	for _, s := range pod.Status.ContainerStatuses {
-		taskRun.Status.Steps = append(taskRun.Status.Steps, v1alpha1.StepState{
-			ContainerState: *s.State.DeepCopy(),
-		})
+		if strings.HasPrefix(s.Name, resources.ContainerPrefix) {
+			taskRun.Status.Steps = append(taskRun.Status.Steps, v1alpha1.StepState{
+				ContainerState: *s.State.DeepCopy(),
+			})
+			if s.State.Terminated == nil {
+				completed = false
+			} else {
+				failure = failure || s.State.Terminated.ExitCode != 0
+			}
+		} else if s.State.Running != nil && s.Ready {
+			sidecarsReady++
+		}
 	}
 
-	switch pod.Status.Phase {
-	case corev1.PodRunning:
-		taskRun.Status.SetCondition(&duckv1alpha1.Condition{
-			Type:   duckv1alpha1.ConditionSucceeded,
-			Status: corev1.ConditionUnknown,
-			Reason: "Building",
-		})
-	case corev1.PodFailed:
-		msg := getFailureMessage(pod)
-		taskRun.Status.SetCondition(&duckv1alpha1.Condition{
-			Type:    duckv1alpha1.ConditionSucceeded,
-			Status:  corev1.ConditionFalse,
-			Message: msg,
-		})
-	case corev1.PodPending:
-		msg := getWaitingMessage(pod)
-		taskRun.Status.SetCondition(&duckv1alpha1.Condition{
-			Type:    duckv1alpha1.ConditionSucceeded,
-			Status:  corev1.ConditionUnknown,
-			Reason:  "Pending",
-			Message: msg,
-		})
-	case corev1.PodSucceeded:
-		taskRun.Status.SetCondition(&duckv1alpha1.Condition{
-			Type:   duckv1alpha1.ConditionSucceeded,
-			Status: corev1.ConditionTrue,
-		})
+	completed = completed || pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed
+	failure = failure || pod.Status.Phase == corev1.PodFailed
+
+	if completed {
+		if failure {
+			msg := getFailureMessage(pod)
+			taskRun.Status.SetCondition(&duckv1alpha1.Condition{
+				Type:    duckv1alpha1.ConditionSucceeded,
+				Status:  corev1.ConditionFalse,
+				Message: msg,
+			})
+		} else {
+			taskRun.Status.SetCondition(&duckv1alpha1.Condition{
+				Type:   duckv1alpha1.ConditionSucceeded,
+				Status: corev1.ConditionTrue,
+			})
+		}
+	} else {
+		switch pod.Status.Phase {
+		case corev1.PodRunning:
+			taskRun.Status.SetCondition(&duckv1alpha1.Condition{
+				Type:   duckv1alpha1.ConditionSucceeded,
+				Status: corev1.ConditionUnknown,
+				Reason: "Building",
+			})
+		case corev1.PodPending:
+			msg := getWaitingMessage(pod)
+			taskRun.Status.SetCondition(&duckv1alpha1.Condition{
+				Type:    duckv1alpha1.ConditionSucceeded,
+				Status:  corev1.ConditionUnknown,
+				Reason:  "Pending",
+				Message: msg,
+			})
+		}
 	}
+
+	sidecars := len(pod.Status.ContainerStatuses) - len(taskRun.Status.Steps)
+	if pod.Status.Phase == corev1.PodRunning && sidecarsReady == sidecars {
```

Comment:
do you think we should worry at all about a state where a sidecar container's status indicates it is "ready", but the underlying binary isn't actually ready? (e.g. the binary is _running_ but not actually ready to handle requests)

Reply:
This approach effectively defines an API/contract for users to follow to signal to tekton that their sidecars are ready, that is, it's defined by the readiness probe of the sidecar. I think that if a use-case falls outside of the ability to be defined by a readiness probe, which can be arbitrary commands/scripts, then it's probably too complex for there to be a way to specify it inside of tekton.

Within user's steps they're free to implement any additional logic they need for more complex ready checks, this really only impacts sidecars which affect automatically specified steps (e.g. git-init). 

I think we wait until there's a clear use-case that readiness probes don't satisfy before coming up with additional ways for user's to specify readiness.

Reply:
Makes sense - sounds great :D